### DISCUSSION TITLE
API compatibililty of the library

### DISCUSSION THREAD

**Discussion by rdeterre** [nlohmann/json#2472] [Q&A] [answered]
Hi,

We're using JSON types from your library in in the APIs of a few different modules (libraries, applications) in our code base. We're using [Conan](https://conan.io/) to manage dependencies and install your library.

Recently, we've noticed that we cannot link modules that use v3.9.1 of the JSON library with other modules that were built against v3.7.3; there have been some API-incompatible changes between the two. This was not flagged by Conan because its default [versioning schema](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#versioning-schema) is to assume that API breaking changes only happen between major releases.

I'd like to propose a change to the nlohmann_json recipe in [conan-center-index](https://github.com/conan-io/conan-center-index) to update the versioning schema used and ensure that other people don't run into the same issue we did, but I couldn't find information regarding API stability in the documentation.

Should we assume that compatibility can be broken in major and minor releases for example, or in any release, including patches?

Thanks,

Romain

**Comment by nlohmann** [comment, marked as answer]
The versioning scheme does not take the ABI into account. A non-breaking change from the point of the library means that existing code can compile after updating the library. We cannot guarantee binary compatibility.

**Reply by rdeterre** [reply]
Thanks, I was confused indeed, it's really the ABI change that we see here. I'll see if this is something that could (and should) be updated in the Conan recipe.
